hearpreprocess/pipeline.py

```diff
@@ -390,98 +382,103 @@ def run(self):
         self.mark_complete()
 
 
-class SubsampleSplit(WorkTask):
+class MetadataTask(WorkTask):
+    """
+    Abstract WorkTask that wants to have access to the metadata
+    from the entire dataset.
+    """
+
+    metadata_task: ExtractMetadata = luigi.TaskParameter()
+    _metadata: Optional[pd.DataFrame] = None
+
+    @property
+    def metadata(self):
+        if self._metadata is None:
+            self._metadata = pd.read_csv(
+                self.metadata_task.workdir.joinpath(self.metadata_task.outfile)
+            )
+        return self._metadata
+
+
+class SubsampleSplit(MetadataTask):
     """
-    A subsampler that acts on a specific split.
-    All instances of this will depend on the combined process metadata csv.
+    For large datasets, we may want to restrict each split to a
+    certain number of minutes.
+    This subsampler acts on a specific split, and ensures we are under
+    our desired audio length threshold for this split.
 
     Parameters:
         split: name of the split for which subsampling has to be done
-        max_files: maximum files required from the subsampling
-        metadata (ExtractMetadata): task which extracts corpus level metadata
-    Requirements:
-        metadata (ExtractMetadata): task which extracts corpus level metadata
     """
 
     split = luigi.Parameter()
-    metadata = luigi.TaskParameter()
 
     def requires(self):
-        # The meta files contain the path of the files in the data
-        # so we dont need to pass the extract as a dependency here.
         return {
-            "metadata": self.metadata,
+            "metadata": self.metadata_task,
         }
 
-    def get_metadata(self):
-        metadata = pd.read_csv(
-            self.requires()["metadata"].workdir.joinpath(
-                self.requires()["metadata"].outfile
-            )
-        )[["split", "split_key", "subsample_key", "slug", "relpath"]]
-
-        # Since event detection metadata will have duplicates, we de-dup
-        # TODO: We might consider different choices of subset
-        metadata = (
-            metadata.sort_values(by="subsample_key")
-            # Drop duplicates as the subsample key is expected to be unique
-            .drop_duplicates(subset="subsample_key", ignore_index=True)
-            # Select the split to subsample
-            .loc[lambda df: df["split"] == self.split]
-        )
-        return metadata
-
     def run(self):
-        metadata = self.get_metadata()
-        num_files = len(metadata)
+        split_metadata = self.metadata[self.metadata["split"] == self.split]
+        relpaths = split_metadata["relpath"].unique()
```

Comment:
Maybe we can use the slug to do this sampling, so it is unique independent of the base directory. Also, the subsampling is dependent on the set of audio in the data. If we want this behaviour it is fine, but we can always use the method used in the splitting(train, test, val). i.e. by hashing followed by the division by 100 and selection.

Reply:
https://github.com/neuralaudio/hear-preprocess/pull/13